internal/archive/archive_test.go

```diff
@@ -371,11 +374,70 @@ func (s *httpSuite) TestArchiveLabels(c *C) {
 	}
 
 	_, err = archive.Open(&options)
+	c.Assert(err, ErrorMatches, `corrupted archive InRelease file: no Ubuntu section`)
+}
+
+func (s *httpSuite) TestProArchives(c *C) {
+	setLabel := func(label string) func(*testarchive.Release) {
+		return func(r *testarchive.Release) {
+			r.Label = label
+		}
+	}
+
+	credsDir := c.MkDir()
+	restore := fakeEnv("CHISEL_AUTH_DIR", credsDir)
+	defer restore()
+
+	confFile := filepath.Join(credsDir, "credentials")
+	contents := ""
+	for _, info := range archive.ProArchiveInfo {
+		contents += fmt.Sprintf("machine %s login foo password bar\n", info.BaseURL)
+	}
+	err := os.WriteFile(confFile, []byte(contents), 0600)
 	c.Assert(err, IsNil)
 
-	s.prepareArchiveAdjustRelease("jammy", "22.04", "amd64", []string{"main", "universe"}, setLabel("ThirdParty"))
+	do := func(req *http.Request) (*http.Response, error) {
+		auth, ok := req.Header["Authorization"]
+		c.Assert(ok, Equals, true)
+		c.Assert(auth, DeepEquals, []string{"Basic Zm9vOmJhcg=="})
+		return s.Do(req)
+	}
+	restoreDo := archive.FakeDo(do)
+	defer restoreDo()
 
-	options = archive.Options{
+	for pro, info := range archive.ProArchiveInfo {
+		s.base = info.BaseURL
+		s.prepareArchiveAdjustRelease("jammy", "22.04", "amd64", []string{"main", "universe"}, setLabel(info.Label))
+
+		options := archive.Options{
+			Label:      "ubuntu",
+			Version:    "22.04",
+			Arch:       "amd64",
+			Suites:     []string{"jammy"},
+			Components: []string{"main", "universe"},
+			CacheDir:   c.MkDir(),
+			Pro:        pro,
+			PubKeys:    []*packet.PublicKey{s.pubKey},
+		}
+
+		_, err = archive.Open(&options)
+		c.Assert(err, IsNil)
+	}
+}
+
+func (s *httpSuite) TestNonProArchive(c *C) {
+	do := func(req *http.Request) (*http.Response, error) {
```

Comment:
I don’t think I understand. I am testing the credentials with a custom `do` function in `TestProArchives`.

```go
	do := func(req *http.Request) (*http.Response, error) {
		auth, ok := req.Header["Authorization"]
		c.Assert(ok, Equals, true)
		c.Assert(auth, DeepEquals, []string{"Basic Zm9vOmJhcg=="})
		return s.Do(req)
	}
	restoreDo := archive.FakeDo(do)
	defer restoreDo()

	for pro, info := range archive.ProArchiveInfo {
		s.base = info.BaseURL
		s.prepareArchiveAdjustRelease("jammy", "22.04", "amd64", []string{"main", "universe"}, setLabel(info.Label))
```

Reply:
In the pro tests, we have the credentials and we are checking that the code is sending them. In this non-pro tests we currently **do not** have the credentials and we are checking the Chisel doesn't send them (but it doesn't have any in the first place). I am saying let's do the same setup that we do for `TestProArchives` and set up the credentials so that we can check that Chisel, with credentials present, does not send them.

My second suggestion was to combine the two tests into one by overriding Do a second time in TestProArchives and then opening the non-pro one and seeing that header is not being sent. This is even better because then the setup is identical and the test is guaranteed to check what it needs to check.

Reply:
Ah oka, gotcha. Done in dd75c92.